File: Jump Game II - Leetcode 45/Jump Game II - Leetcode 45.py
# Brute Force
class Solution:
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        smallest = [float('inf')]

        def backtrack(i=0, jumps=0):
            if i == n-1:
                smallest[0] = min(smallest[0], jumps)
                return
            
            max_jump = nums[i]
            max_reachable_index = min(i+max_jump, n-1)

            for new_index in range(max_reachable_index, i, -1):
                backtrack(new_index, jumps+1)
        
        backtrack()
        return smallest[0]
    
    
# Optimal    
class Solution:
    def jump(self, nums: List[int]) -> int:
        smallest = 0
        n = len(nums)
        end, far = 0, 0

        for i in range(n-1):
            far = max(far, i+nums[i])
            
            if i == end:
                smallest += 1
                end = far
        
        return smallest
        # Time: O(n)
        # Space: O(1)
        

File: Jump Game II - Leetcode 45/Jump Game II - Leetcode 45.cpp
// Brute Force
#include <vector>
#include <algorithm>
#include <climits>

class Solution {
public:
    int jump(std::vector<int>& nums) {
        int n = nums.size();
        int smallest = INT_MAX;

        backtrack(0, 0, nums, n, smallest);
        return smallest;
    }

private:
    void backtrack(int i, int jumps, std::vector<int>& nums, int n, int& smallest) {
        if (i == n - 1) {
            smallest = std::min(smallest, jumps);
            return;
        }

        int maxJump = nums[i];
        int maxReachableIndex = std::min(i + maxJump, n - 1);

        for (int newIndex = maxReachableIndex; newIndex > i; newIndex--) {
            backtrack(newIndex, jumps + 1, nums, n, smallest);
        }
    }
};


// Optimal
#include <vector>
#include <algorithm>

class Solution {
public:
    int jump(std::vector<int>& nums) {
        int smallest = 0;
        int n = nums.size();
        int end = 0;
        int far = 0;

        for (int i = 0; i < n - 1; i++) {
            far = std::max(far, i + nums[i]);
            
            if (i == end) {
                smallest++;
                end = far;
            }
        }

        return smallest;
    }
};
